packages/headless/doc-parser/src/extractor.ts

```diff
@@ -0,0 +1,59 @@
+import {AnyEntity, isObjectEntity, ObjEntity} from './entity';
+import {buildEntity, buildObjEntity} from './entity-builder';
+
+interface Extraction {
+  types: ObjEntity[];
+}
+
+export function extractTypes(headingEntities: AnyEntity[]): Extraction {
+  const extraction: Extraction = {
+    types: [],
+  };
+
+  processObjectEntities(headingEntities, extraction, 0);
+
+  return extraction;
+}
+
+function processObjectEntities(
+  entities: AnyEntity[],
+  extraction: Extraction,
+  level: number
+) {
+  entities
+    .filter(isObjectEntity)
+    .forEach((entity) => extract(entity, extraction, level + 1));
+}
+
+function extract(entity: ObjEntity, extraction: Extraction, level: number) {
+  const maxDepth = 2;
+  const {members} = entity;
+
+  if (!members.length) {
+    return;
+  }
+
+  if (level === maxDepth) {
+    const typeHeading = buildTypeEntity(entity, members);
+    extraction.types.push(typeHeading);
```

Comment:
Before pushing a type heading to the extracted types list, perhaps we should check if there's already a type heading with the same `name` and different properties? Unfortunately, I feel like we'll occasionally accidentally have two interfaces with the same type name (such as response interface vs redefined interfaces in the controller), and throwing an error here could help us correct that early on.

To go along with that, perhaps the type of `extraction.types` could be changed to a record such as `{ [name: string]: ObjEntity }` or `Record<string, ObjEntity>`, which would make this check easier and reflect that there can't be two ObjEntity with the same name.

Reply:
> Unfortunately, I feel like we'll occasionally accidentally have two interfaces with the same type name

Good point, but maybe one clarification is that the extraction happens on a per controller level, not on all headless controllers. I don't think it would be possible to have duplicate names with different properties because it would require two different interfaces sharing the same name in a single headless file, which is something typescript will flag.

During my tests on the facet, I did not see this happen, but if it does then we will need to adjust some things,